File: C++/Command/SmartHome/command.cc
#include <iostream>
#include "command.h"

LightOnCommand::LightOnCommand(const std::shared_ptr<Light> &light) : light(light) {}

void LightOnCommand::execute() {
    light->on();
}

LightOffCommand::LightOffCommand(const std::shared_ptr<Light> &light) : light(light) {}

void LightOffCommand::execute() {
    light->off();
}


TVOnCommand::TVOnCommand(const std::shared_ptr<TV> &tv) : tv(tv) {}

void TVOnCommand::execute() {
    tv->on();
    switch_channel();
}

void TVOnCommand::switch_channel() {
    int channel_number;
    std::cout << "Choose a channel...\n";
    std::cin >> channel_number;

    if (std::cin.fail()) {
        std::cout << "Invalid channel number...\nSetting channel number to 0...\n";
        channel_number = 0;
    }

    tv->switch_channel(channel_number);
}

TVOffCommand::TVOffCommand(const std::shared_ptr<TV> &tv) : tv(tv) {}

void TVOffCommand::execute() {
    tv->off();
}

File: Java/Command/SmartHome/src/com/ericgtkb/Command.java
package com.ericgtkb;

import java.util.Scanner;

public interface Command {
    void execute();
}


class LightOnCommand implements Command {
    private Light light;

    public LightOnCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.on();
    }
}


class LightOffCommand implements Command {
    private Light light;

    public LightOffCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.off();
    }
}


class TVOnCommand implements Command {
    private TV tv;

    public TVOnCommand(TV tv) {
        this.tv = tv;
    }

    @Override
    public void execute() {
        tv.on();
        switchChannel();
    }

    private void switchChannel() {
        int channelNumber;

        System.out.println("Choose a channel...");
        try (Scanner scanner = new Scanner(System.in)) {
            channelNumber = scanner.nextInt();
        } catch (Exception e) {
            System.out.println("Invalid channel number...\nSetting channel number to 0...");
            channelNumber = 0;
        }

        tv.switchChannel(channelNumber);
    }
}


class TVOffCommand implements Command {
    private TV tv;

    public TVOffCommand(TV tv) {
        this.tv = tv;
    }

    @Override
    public void execute() {
        tv.off();
    }
}


